### PR TITLE
Add flag to enable/disable dataCache in ClusterStateCache

### LINKED ISSUE #23994: [Feature] [IM] Create the dedicated cache backed by ReadClient::Callback for highest event number and cluster version
### Feature description

```bash
For data version and event number, currently it is set via update cache in cluster state cache when ReadClient::Callback's onAttributeData and onEventData is called, and is pretty useful to reduce traffic in resubscription user case. For event number, assume we have upgraded callback backed by data cache, yes, it would be fairly easy to store event number in new cache. But for data version, it is more related to cluster data itself, in other works, each cluster's data matches with one data version, it is more intuitive to put cluster version along with data together in one cache. Definitely it is technically possible that we have two cache, one cache for attribute/event data, and another cache for event number and cluster data version, we just copy a cluster state cache then delete data logic inside, then add new clusterEventVersionCache member inside cluster state  cache, then clusterEventVersionCache has BufferedReadCallback member inside, this way, existing logic would not be broken, further you can use new clusterEventVersionCache  directly, I guess the benefit to refactor the "event number and data version" bits out of cluster state could be that folks may wanna use event number and data version from "clusterEventVersionCache" but don't wanna clusterDataCache in subscription case, in that situation, those folk may already have their dedicated data cache, and don't wanna the stl c++ map cache.....for example, kotlin, swift developer may wanna get their cache storage in their language for more flexibility.
```

### Platform

all

### Platform Version(s)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by yunhanw-google**
https://github.com/project-chip/connectedhomeip/issues/23994
Add flag to enable/disable dataCache in ClusterStateCache and add the tests for event/data version cache without storing the attribute/event data.

### CODE REVIEW

File: src/app/ClusterStateCache.h [hunk 1]
```diff
@@ -86,8 +86,10 @@ class ClusterStateCache : protected ReadClient::Callback
         virtual void OnEndpointAdded(ClusterStateCache * cache, EndpointId endpointId){};
     };
 
-    ClusterStateCache(Callback & callback, Optional<EventNumber> highestReceivedEventNumber = Optional<EventNumber>::Missing()) :
-        mCallback(callback), mBufferedReader(*this)
+    ClusterStateCache(Callback & callback, bool cacheData = true,
```
Comment (bzbarsky-apple) [inline src/app/ClusterStateCache.h:89]: This should document what passing `cacheData = false` would mean.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/ClusterStateCache.h (hunk 1 above)
Comment (jtung-apple) [inline src/app/ClusterStateCache.h:89]: Please consider moving the new argument with default value to the end of argument list.

Comment (yunhanw-google) [reply src/app/ClusterStateCache.h:89]: done, thanks

Comment (yunhanw-google) [reply src/app/ClusterStateCache.h:89]: done, thanks

**Review by yufengwangca — APPROVED** [review]

**Pull request merged**